allium/include/Parser/Parser.h

```diff
@@ -7,43 +7,141 @@
 #include "Parser/Lexer.h"
 #include "Utils/Optional.h"
 #include "Utils/TaggedUnion.h"
+#include "Utils/ParserValues.h"
 
 namespace parser {
 
+/// Custom error class representing any syntax errors encountered during parsing.
+/// The parser terminates and returns an empty AST on the first syntax error it encounters.
+class SyntaxError {
+public:
+    SyntaxError(const std::string& msg, SourceLocation location) : msg(msg), location(location) {}
+    ~SyntaxError() {}
+
+    friend bool operator==(const SyntaxError &lhs, const SyntaxError &rhs) {
+        return lhs.msg == rhs.msg && lhs.location == rhs.location;
+    }
+
+    friend std::ostream& operator<<(std::ostream& stream, const SyntaxError& error) {
+        stream << "syntax error " << error.getLocation() << " - " << error.getMessage() << std::endl;
+        return stream;
+    }
+
+    std::string getMessage() const {return(msg);};
+    SourceLocation getLocation() const {return(location);};
+private:
+    /// The error message associated with the error
+    std::string msg;
+
+    /// The location of the error in the source file.
+    SourceLocation location;
+};
+
+// Class representing either an optional or a list of errors for a particular parse method.
+template <typename T>
+class ParserResult: public TaggedUnion<Optional<T>, std::vector<SyntaxError>> {
```

Comment:
Judging by the constructors and a few of the parsers you've written, an empty error list and a valueless optional mean the same thing — parsing failed without an error (e.g. `parseTruthLiteral()`). It seems you've modeled it this way:
> A `ParserResult<T>` is either an optional value of type T or a non-empty list of syntax errors.

because a list can be empty, and an optional can have no value, that model is denormalized (there is a value with more than one representation). I claim this model is normalized:
> A `ParserResult<T>` is either a value of type T or a (possibly empty) list of syntax errors.

In addition to the benefits of normalized representations, this also makes it possible to access the value of a result in a single step, instead of first accessing the optional value and then unwrapping it.

I'm only requesting a change in the representation — you can keep the constructor that takes an optional value and a  list of errors that you use in many of the parsers, you'll just need to tweak the constructor body a bit.

Reply:
Thanks for this suggestion - the ParserResult representation was bothering me before but I couldn't put my finger on why. Cleaned up the representation in ecbdc468d95a737fd22ed763aa106883903c2476. I ended up removing the references to optionals from the parser altogether because there was ultimately no reason to keep them outside of the context of constructing a ParserResult.